Enhancement to move to OpenAPI Spec 3.0.0

abellotti opened the pull request:
Enhancement to update the API to:
- move from swagger 2 to openapi 3.0.0
- rename the swagger:generate task to  openapi:generate and update accordingly.

Updating OpenAPI 3.0 specs now done via:
  $ bundle exec rake openapi:generate

abellotti commented:
Marking as WIP, holding place for this work until we're ready to pull the trigger.

Fryguy commented on lib/api/docs.rb:
```diff
@@ -1,3 +1,4 @@
 module Api
-  Docs = ::OpenApi::Docs.new(Dir.glob(Rails.root.join("public/doc/swagger*.yaml")))
+  Docs = ::OpenApi::Docs.new(Dir.glob(Rails.root.join("public/doc/swagger*.yaml")) +
+                             Dir.glob(Rails.root.join("public/doc/openapi*.yaml")))
```
In catalog, and in the IPP, I believe we want to go to .json format for the new openapi file.

abellotti replied:
yep, cleaner too. moved to openapi*.json for the OpenAPI 3.x specifications.

Fryguy commented on lib/tasks/openapi_generate.rake:
```diff
@@ -0,0 +1,388 @@
+  end
+
+  def openapi_file
+    Pathname.new(__dir__).join("../../public/doc/openapi-3-v0.2.0.yaml").to_s
```
This feels weird that we would hardcode the 0.2.0 version in the generator...I'd expect it as a parameter or something.

abellotti replied:
Agreed. version is now generated from the latest openapi.json route defined for the project.

Fryguy commented on lib/tasks/openapi_generate.rake:
```diff
@@ -0,0 +1,388 @@
+class OpenapiGenerator
```
I see most of this is just a copy of the existing swagger generator.  I'm concerned about duplication, but that can be cleaned up in a follow up.  Even so, can you show a diff between the two so I can figure out how to review this?

abellotti replied:
I think this is now better, I've essentially git mv'd the original swagger_generate.rake to the new openapi_generate.rake successor, so git diffs now show the changes.

Fryguy commented on lib/api/docs.rb:
```diff
@@ -1,3 +1,4 @@
 module Api
-  Docs = ::OpenApi::Docs.new(Dir.glob(Rails.root.join("public/doc/swagger*.yaml")))
+  Docs = ::OpenApi::Docs.new(Dir.glob(Rails.root.join("public", "doc", "openapi*.json")) +
+                             Dir.glob(Rails.root.join("public", "doc", "swagger*.yaml")))
```
Can we just drop the yamls?

Fryguy replied:
Or put another way, can we just drop v2 altogether, since we are going to cut a 1.0 anyway

abellotti replied:
We can drop the yamls & v2 Doc library since we don't expose /api/v0.0/openapi.json anyways,  want me to do in this PR ?

Fryguy replied:
yup 👍

Fryguy commented on app/controllers/api/v0x1.rb:
```diff
@@ -2,7 +2,7 @@ module Api
     class RootController < ApplicationController
       def openapi
-        render :json => Api::Docs["0.1"].to_json
+        render :json => Api::Docs["0.1"].content.to_json
```
This change is not clear to me.  Why doesn't the to_json itself just call content?

abellotti replied:
it does, it's just that the spec should not include the content key which is what to_json serializes it as:

```
{
  "content" : {
    "openapi" : "3.0.0",
    "info" : { ... },
    "security" : { ... },
    "paths" : { 
       ...
    }
    ...
}
```

should just be:

```
{
  "openapi" : "3.0.0",
  "info" : { ... },
  "security" : { ... },
  "paths" : {
     ...
  }
  ...
}
```

same issue with the current v0.1.0 swagger flavor.

Fryguy commented on lib/open_api/docs/component_collection.rb:
```diff
@@ -13,7 +13,7 @@ def [](name)
 
       def load_definition(name)
-        raw_definition = @doc.content[@category][name]
+        raw_definition = @doc.content.fetch_path(@category.split("/"))[name]
```
If you're using fetch_path, you can put the name in the fetch as well

```ruby
@doc.content.fetch_path(*@category.split("/"), name)
```

abellotti replied:
👍

Fryguy commented on lib/open_api/docs.rb:
```diff
@@ -7,7 +7,11 @@ def initialize(glob)
+              DocV2.new(yaml)
+            elsif yaml["openapi"] =~ /3*/
+              DocV3.new(yaml)
+            end
```
Not sure I follow the need for this if we don't have v2 anymore.  Additionally, it's not a yaml anymore, right, so is the variable name wrong?

abellotti replied:
yaml really just a hash :) this code would change of course if we drop DocV2 altogether.

abellotti commented:
Ok @Fryguy, I've removed the DocV2 code and simplified what we had in this PR.  I've also converted the v0.0.1 yaml to OpenAPI 3.0.0 json as this would be needed for the /internal/v0 paths anyways.  It also allows us to remove /api/v0.0 in a separate PR. let me know if you'd like to see other changes in this PR. Thank you.

Ladas commented on lib/tasks/openapi_generate.rake:
```diff
@@ -0,0 +1,416 @@
+      when :boolean
+        properties_value["type"] = "boolean"
+      when :jsonb
+        ['type', 'items'].each do |property_key|
```
hmm you're missing this line https://github.com/ManageIQ/topological_inventory-api/blob/b9c30161c95b8fef9d85a002674ab04b7d796c3c/lib/tasks/swagger_generate.rake#L177

it was probably added after you took the generator?

abellotti replied:
👍  Good catch!!  Thanks @Ladas for 👀 at this.  Will push an update.

Fryguy commented on lib/open_api/docs.rb:
```diff
@@ -6,8 +6,8 @@ def initialize(glob)
-      yaml = YAML.load_file(file)
-      doc  = DocV2.new(yaml) if yaml["swagger"] == "2.0"
+      openapi_spec = JSON.parse(File.read(file))
+      doc = DocV3.new(openapi_spec) if openapi_spec["openapi"] =~ /3*/
```
if it doesn't match `/3*/` what happens?  Should it raise?

Fryguy commented on lib/open_api/docs/doc_v3.rb:
```diff
@@ -11,37 +11,34 @@ def version
 
       def base_path
-        @base_path ||= @content["basePath"]
+        @base_path ||= @content["servers"].first["variables"]["basePath"]["default"]
```
Minor...

```ruby
@content.fetch_path("servers, 0, "variables", "basePath", "default")
```

Fryguy commented on lib/open_api/docs/doc_v3.rb:
```diff
@@ -11,37 +11,34 @@ def version
-            end
-          end
-        end
+        @routes ||= Docs.path_routes(base_path, paths)
```
I don't think you need this refactor anymore...you did before when you were trying to share v2 and v3 code, but now you don't.

abellotti commented:
Ping @Fryguy, all changes made and pushed out. Anything else needed here ?

mkanoor commented:
@abellotti 
Line 782
      operationId: showServiceOfferingIcon
Line 798
      operationId: showServiceOfferingIcon
They have to be unique, when you load the son file in the swagger editor it complains

abellotti commented:
I know, let's fix that after this PR is merged since it's a current issue with the swagger too.  Essentially the openapi.generate needs to be enhanced to know about the case where we're getting a single subresource (and not listing), so the operationId generated for this would be something like: "showServiceOfferingIconIconData". Thanks.
